primitive/integers.go

```diff
@@ -52,13 +52,27 @@ func ReadShort(source io.Reader) (decoded uint16, err error) {
 	return decoded, err
 }
 
+func ReadShortSigned(source io.Reader) (decoded int16, err error) {
```

Comment:
Ì'm not a fan of the additions in this PR to the `primitive` package. In principle this package is meant to map protocol structures defined in the specs, section 3. The `[short]` type is defined as "2 bytes unsigned integer" in all protocol versions that I'm aware of. In go this is `uint16`.

Reply:
I think I was too much focused to mimic V4 protocol. All protocols define short as `[short] A 2 bytes unsigned integer` and `[int] A 4 bytes integer`. When it comes to collections and map length things go differently:

V2:
> List: a [short] n indicating the size of the list, followed by n elements.
  Each element is [short bytes] representing the serialized element
  value.

V4:
> 6.11 list
  A [int] n indicating the number of elements in the list, followed by n
  elements.  Each element is [bytes] representing the serialized value.

Strictly speaking V2 short does not allow negative values which was the convention in V4 to indicate NULL collection elements. I have updated PR to strictly follow V2 specification. I had to adopt tests unfortunately, because if we do not store `-1` as NULL collection element length, we cannot distinguish between NULL and empty values (e.g. `nil` vs `""` strings).

Java driver itself does not support NULL values in collections and maps. Gocql does not support protocol V2 version, so it is hard to manually tests this specific use-case.

Reply:
> Gocql does not support protocol V2 version, so it is hard to manually tests this specific use-case.

We should be able to reproduce whatever test case we want with what is being used on this library to test things, no matter if it's the "local" client and server or just unit tests in general.

So, atm what happens if you "ask" the library to encode a `nil` collection? It will encode it as if it was an empty `collection`? I wonder if it should return an error in that case...

Reply:
Encoding `nil` value in collection will result in the same as encoding _empty_ value. We cannot write `-1` as length and it will be zero.

After further thought, I think about raising error if `nil` is passed for C* 2.0.0, as those requests will not be accepted by the server:
```
Connected to test1 at 127.0.0.1:9160.
[cqlsh 4.0.0 | Cassandra 2.0.0 | CQL spec 3.1.0 | Thrift protocol 19.37.0]
Use HELP for help.
cqlsh> INSERT INTO test.my_list(k, m) VALUES(10, ['1', NULL, '2']);
Bad Request: null is not supported inside collections
cqlsh> INSERT INTO test.my_set(k, m) VALUES(10, {'1', NULL, '2'}) ;
Bad Request: null is not supported inside collections
```